control/HTTPCacheControl.php

```diff
@@ -0,0 +1,357 @@
+<?php
+
+/**
+ * Class HTTPCacheControl
+ *
+ *
+ */
+class HTTPCacheControl extends SS_Object {
+
+	/**
+	 * @var static
+	 */
+	private static $inst;
+
+	/**
+	 * Store for all the current directives and their values
+	 *
+	 * @var array
+	 */
+	private $state = array();
+
+	/**
+	 * Whether the cache-control object is locked to further changes
+	 *
+	 * @var bool
+	 */
+	protected $locked = false;
+
+	/**
+	 * A list of allowed cache directives for HTTPResponses
+	 *
+	 * This doesn't include any experimental directives,
+	 * use the config system to add to these if you want to enable them
+	 *
+	 * @config
+	 * @var array
+	 */
+	private static $allowed_directives = array(
+		'public',
+		'private',
+		'no-cache',
+		'max-age',
+		's-maxage',
+		'must-revalidate',
+		'proxy-revalidate',
+		'no-store',
+		'no-transform',
+	);
+
+	/**
+	 * Lock the current state of the cache control to prevent further modifications
+	 *
+	 * @return $this
+	 */
+	public function lock()
+	{
+		$this->locked = true;
+		return $this;
+	}
+
+	/**
+	 * Low level method for setting directives include any experimental or custom ones added via config
+	 *
+	 * @param string $directive
+	 * @param string|bool $value
+	 *
+	 * @return $this
+	 */
+	public function setDirective($directive, $value = null)
+	{
+		if ($this->locked) {
```

Comment:
I think this should only complain if you're trying to set a *different* key/value than what's currently locked. I can see situations where the system calls `disableCaching()` multiple times in some execution flows, and that's technically not an error.